src/Optimization/hiopDualsUpdater.cpp

```diff
@@ -46,57 +46,106 @@
 // product endorsement purposes.
 
+/**
+ * @file hiopDualsUpdater.cpp
+ *
+ * @author Cosmin G. Petra <petra1@llnl.gov>,  LLNL
+ * @author Nai-Yuan Chiang <chiang7@llnl.gov>,  LLNL
+ *
+ */
+ 
 #include "hiopDualsUpdater.hpp"
 #include "hiopLinAlgFactory.hpp"
 
 #include "hiop_blasdefs.hpp"
 
+#ifdef HIOP_SPARSE
+#ifdef HIOP_USE_COINHSL
+#include "hiopLinSolverIndefSparseMA57.hpp"
+#endif
+#ifdef HIOP_USE_STRUMPACK
+#include "hiopLinSolverSparseSTRUMPACK.hpp"
+#endif
+#endif
+
 namespace hiop
 {
 
 hiopDualsLsqUpdate::hiopDualsLsqUpdate(hiopNlpFormulation* nlp) 
-  : hiopDualsUpdater(nlp) 
+  : hiopDualsUpdater(nlp), 
+    _mexme_(nullptr), _mexmi_(nullptr), _mixmi_(nullptr), _mxm_(nullptr), 
+    M_(nullptr),
+    rhs_(nullptr), rhsc_(nullptr), rhsd_(nullptr),
+    _vec_n_(nullptr),_vec_mi_(nullptr),
+    linSys_(nullptr),
+    lsq_dual_init_max(1e3),
+    augsys_update_(false) 
 {
-  hiopNlpDenseConstraints* nlpd = dynamic_cast<hiopNlpDenseConstraints*>(_nlp);
-  assert(NULL!=nlpd);
-  _mexme = LinearAlgebraFactory::createMatrixDense(nlpd->m_eq(),   nlpd->m_eq());
-  _mexmi = LinearAlgebraFactory::createMatrixDense(nlpd->m_eq(),   nlpd->m_ineq());
-  _mixmi = LinearAlgebraFactory::createMatrixDense(nlpd->m_ineq(), nlpd->m_ineq());
-  _mxm   = LinearAlgebraFactory::createMatrixDense(nlpd->m(), nlpd->m());
-
-  M      = LinearAlgebraFactory::createMatrixDense(nlpd->m(), nlpd->m());
-  rhs    = LinearAlgebraFactory::createVector(nlpd->m());
-  rhsc   = nlpd->alloc_dual_eq_vec();
-  rhsc->setToZero();
-  rhsd   = nlpd->alloc_dual_ineq_vec();
-  rhsd->setToZero();
-  _vec_n = nlpd->alloc_primal_vec();
-  _vec_mi= nlpd->alloc_dual_ineq_vec();
+  hiopNlpSparse* nlpsp = dynamic_cast<hiopNlpSparse*>(_nlp);
+  if(nullptr!=nlpsp)
+  {
+#ifndef HIOP_SPARSE
+      assert(0 && "should not reach here!");
+#endif // HIOP_SPARSE        
+      augsys_update_ = true;
+      rhs_    = LinearAlgebraFactory::createVector(_nlp->n() + _nlp->m_ineq() + _nlp->m());
+  }
+  else
+  {
+    _mexme_ = LinearAlgebraFactory::createMatrixDense(_nlp->m_eq(),   _nlp->m_eq());
+    _mexmi_ = LinearAlgebraFactory::createMatrixDense(_nlp->m_eq(),   _nlp->m_ineq());
+    _mixmi_ = LinearAlgebraFactory::createMatrixDense(_nlp->m_ineq(), _nlp->m_ineq());
+    _mxm_   = LinearAlgebraFactory::createMatrixDense(_nlp->m(), _nlp->m());
+
+    M_      = LinearAlgebraFactory::createMatrixDense(_nlp->m(), _nlp->m());
+    rhs_    = LinearAlgebraFactory::createVector(_nlp->m());
 #ifdef HIOP_DEEPCHECKS
-  M_copy = M->alloc_clone();
-  rhs_copy = rhs->alloc_clone();
-  _mixme = LinearAlgebraFactory::createMatrixDense(nlpd->m_ineq(), nlpd->m_eq());
+    M_copy_ = LinearAlgebraFactory::createMatrixDense(_nlp->m(), _nlp->m());
+    rhs_copy_ = LinearAlgebraFactory::createVector(_nlp->m());
+    _mixme_ = LinearAlgebraFactory::createMatrixDense(_nlp->m_ineq(), _nlp->m_eq());
+#endif
+  }
+
+#ifdef HIOP_USE_MPI
+  long long* vec_distrib = _nlp->get_vec_distrib();
```

Comment:
I find it the following lines a bit too complicated/much. This class should not know about `vec_distrib`. Apparently we have two "factories" for linear algebra, `LinearAlgebraFactory` and `hiopNlpFormulation`. `LinearAlgebraFactory` decides between memory spaces, but it is up to implementations of `hiopNlpFormulation` to decide whether the vectors are MPI-based: `hiopNlpSparse` does not need MPI while `hiopNlpDenseConstraints` does. I looks like it will be cleaner to allocate via `hiopNlpFormulation` `alloc_primal_vector` and other `alloc_xxx` functions and within these functions to use `LinearAlgebraFactory` to instantiate the right mem space type vectors/matrices. @pelesh @ashermancinelli @nychiang what do you think?

Reply:
Yes, I think that's a good idea. Currently we are using `LinearAlgebraFactory::createVector` without MPI information in many places of hiopKKTxxxx classes. It's better to change all of them to  `hiopNlpFormulation::allco_xxx` and let `hiopNlpFormulation` to determine the MPI usage.

Reply:
How cleanly is the distribution logic in `hiopNlpFormulation` implemented? If it's pretty clean, I think it would make more sense to move that logic to `LinearAlgebraFactory`. At this point it might make sense to think about extending linalg factory with additional implementations as we discussed last quarter. I don't want to bring in a bunch of scope creep though, so we can discuss this elsewhere if it would be too much work for this pr.

Reply:
@ashermancinelli  `LinearAlgebraFactory` already knows how to create MPI vectors of all sorts. And `hiopNlpFormulation` uses internally that functionality from `LinearAlgebraFactory` -- if/when needed, depending on the nlp formulation type. What I am saying is that in hiopKKTxxxx classes, we should use `hiopNlpFormulation` to do the allocations.

Reply:
I see, this makes perfect sense. Sounds great to me.